src/runtime_src/core/common/api/xrt_hw_context.cpp

```diff
@@ -207,6 +276,41 @@ alloc_hwctx_from_mode(const xrt::device& device, const xrt::uuid& xclbin_id, xrt
   return handle;
 }
 
+static std::shared_ptr<hw_context_impl>
+alloc_empty_hwctx(const xrt::device& device, const xrt::hw_context::cfg_param_type& cfg_param, xrt::hw_context::access_mode mode)
+{
+  XRT_TRACE_POINT_SCOPE(xrt_hw_context);
+  auto handle = std::make_shared<hw_context_impl>(device.get_handle(), cfg_param, mode);
+
+  // Update device is called with a raw pointer to dyanamically
+  // link to callbacks that exist in XDP via a C-style interface
+  // The create_hw_context_from_implementation function is then 
+  // called in XDP create a hw_context to the underlying implementation
+  xrt_core::xdp::update_device(handle.get());
+
+  handle->get_usage_logger()->log_hw_ctx_info(handle.get());
+
+  return handle;
+}
+
+static std::shared_ptr<hw_context_impl>
+alloc_hwctx_from_elf(const xrt::device& device, const xrt::elf& elf, const xrt::hw_context::cfg_param_type& cfg_param,
+                     xrt::hw_context::access_mode mode)
+{
+  XRT_TRACE_POINT_SCOPE(xrt_hw_context);
+  auto handle = std::make_shared<hw_context_impl>(device.get_handle(), elf, cfg_param, mode);
+
+  // Update device is called with a raw pointer to dyanamically
+  // link to callbacks that exist in XDP via a C-style interface
+  // The create_hw_context_from_implementation function is then 
+  // called in XDP create a hw_context to the underlying implementation
+  xrt_core::xdp::update_device(handle.get());
+
+  handle->get_usage_logger()->log_hw_ctx_info(handle.get());
+
+  return handle;
+}
```

Comment:
Little concerning with duplicated code.  How about?


```suggestion
std::shared_ptr<hw_context_impl>
static post_allow_hwctx(const std::shared_ptr<hw_context_impl>& handle)
{
  // Update device is called with a raw pointer to dyanamically
  // link to callbacks that exist in XDP via a C-style interface
  // The create_hw_context_from_implementation function is then 
  // called in XDP create a hw_context to the underlying implementation
  xrt_core::xdp::update_device(handle.get());

  handle->get_usage_logger()->log_hw_ctx_info(handle.get());

  return handle;
}

static std::shared_ptr<hw_context_impl>
alloc_empty_hwctx(const xrt::device& device, const xrt::hw_context::cfg_param_type& cfg_param, xrt::hw_context::access_mode mode)
{
  XRT_TRACE_POINT_SCOPE(xrt_hw_context);
  return post_alloc_hwctx(std::make_shared<hw_context_impl>(device.get_handle(), cfg_param, mode));
}


static std::shared_ptr<hw_context_impl>
alloc_hwctx_from_elf(const xrt::device& device, const xrt::elf& elf, const xrt::hw_context::cfg_param_type& cfg_param,
                     xrt::hw_context::access_mode mode)
{
  XRT_TRACE_POINT_SCOPE(xrt_hw_context);
  return post_alloc_hwctx(std::make_shared<hw_context_impl>(device.get_handle(), elf, cfg_param, mode));
}
```

Reply:
Yes Soren, this removes code duplicity. Thanks